cli/cmd/test_connection.go

```diff
@@ -171,34 +185,146 @@ func testConnection(cmd *cobra.Command, args []string) error {
 	}
-	updateSyncTestConnectionStatus(ctx, log.Logger, status)
+	updateSyncTestConnectionStatus(context.Background(), log.Logger, status, testConnectionResults...)
+
+	log.Info().Any("testresults", testConnectionResults).Msg("Test connection completed")
+
+	maxLength := 0
+	for _, testResult := range testConnectionResults {
+		if len(testResult.pluginRef) > maxLength {
+			maxLength = len(testResult.pluginRef)
+		}
+	}
+
+	connFailures := make([]testConnectionResult, 0, len(testConnectionResults))
+	for i, testResult := range testConnectionResults {
+		if i == 0 {
+			cmd.Println("Test Connection Results")
+		}
+		cmd.Printf("%-12s %s %s", testResult.pluginKind, testResult.pluginRef, strings.Repeat(" ", maxLength-len(testResult.pluginRef)+1))
+		if testResult.Success {
+			cmd.Println("Success")
+			continue
+		}
+		cmd.Println("Failure:", testResult.FailureCode, "\t", testResult.FailureDescription)
+		connFailures = append(connFailures, testConnectionResults[i])
+	}
+
+	if len(connFailures) > 0 {
+		allErrors = errors.Join(allErrors, &testConnectionFailures{failed: connFailures})
+	}
 
 	return allErrors
 }
+
+type testConnectionFailures struct {
+	failed []testConnectionResult
+}
+
+func (*testConnectionFailures) Error() string {
+	// The errs are already shown in the console, so we don't need to show them again here
+	return "at least one test connection failed"
+}
+
+type testConnectionResult struct {
+	pluginRef          string
+	pluginKind         string
+	Success            bool
+	FailureCode        string `json:",omitempty"`
+	FailureDescription string `json:",omitempty"`
+}
+
+func testPluginConnection(ctx context.Context, client plugin.PluginClient, spec map[string]any) (*testConnectionResult, error) {
+	specBytes, err := marshalSpec(spec)
+	if err != nil {
+		return nil, err
+	}
+
+	in := &plugin.TestConnection_Request{
+		Spec: specBytes,
+	}
+
+	resp, err := client.TestConnection(ctx, in)
+	if err != nil {
+		if gRPCErr, ok := grpcstatus.FromError(err); ok {
+			if gRPCErr.Code() == codes.Unimplemented {
+				err := initPlugin(ctx, client, spec, false, invocationUUID.String())
+				if err != nil {
+					return &testConnectionResult{
+						Success:            false,
+						FailureCode:        "OTHER",
+						FailureDescription: err.Error(),
+					}, nil
+				}
+
+				return &testConnectionResult{
+					Success: true,
+				}, nil
+			}
+		}
+		return nil, err
+	}
+
+	return &testConnectionResult{
+		Success:            resp.Success,
+		FailureCode:        resp.FailureCode,
+		FailureDescription: resp.FailureDescription,
+	}, nil
+}
+
+// filterFailedTestResults fetch the failed test results.
+//
+// The function returns any failed test results, or nil if all tests passed. The hackernews plugin is excluded from the
+// failed test results since it was previously used as a test case for the test connection command.
+func filterFailedTestResults(results []testConnectionResult) (*testConnectionResult, error) {
```

Comment:
Wait what's the reason we need to filter these plugins? I don't think their connection test can ever fail

Reply:
Those should be there until we update the cloud side to use the new cli.

Reply:
Ah because with existing Cloud the new CLI will report failures for this?

Reply:
I don't think it matters much this feature is only enabled for the `cloudquery` team

Reply:
If we want the cloud api side to work, this filtering needs to be there at least for now.